fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -238,4 +249,134 @@ private static void setDeprecated(DescriptorProtos.FieldDescriptorProto.Builder
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                if (!hasField(unionDescriptorBuilder, messageType)) {
+                    addFieldToUnion(unionDescriptorBuilder, messageType.getName(), "." + messageType.getFullName());
+                }
+            }
+        }
+        fileBuilder.addMessageType(unionDescriptorBuilder);
+        try {
+            return Descriptors.FileDescriptor.buildFrom(fileBuilder.build(), fileDescriptor.getDependencies().toArray(new Descriptors.FileDescriptor[0]));
+        } catch (Descriptors.DescriptorValidationException e) {
+            throw new MetaDataException("Failed to add a default union", e);
+        }
+    }
+
+    @Nonnull
+    private static DescriptorProtos.DescriptorProto.Builder createDefaultUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor) {
+        DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
+        unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
+        for (Descriptors.Descriptor messageType : fileDescriptor.getMessageTypes()) {
+            RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType.toProto());
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                addFieldToUnion(unionMessageType, messageType.getName(), messageType.getFullName());
+            }
+        }
+        return unionMessageType;
+    }
+
+    /**
+     * Creates a default union descriptor for the given file descriptor and a base union descriptor. It adds all of the
+     * non-{@code NESTED} message types that exist in the base union to the synthetic union.
+     * @param fileDescriptor the file descriptor to create a union for
+     * @param baseUnionDescriptor the base union descriptor
+     * @return the builder for the union
+     */
+    @Nonnull
+    @API(API.Status.INTERNAL)
+    public static DescriptorProtos.DescriptorProto.Builder createSyntheticUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor, @Nonnull Descriptors.Descriptor baseUnionDescriptor) {
+        DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
+        unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
+        if (baseUnionDescriptor.getOneofs().size() > 0) {
+            throw new MetaDataException("Adding record type to oneof is not allowed");
+        }
+        for (Descriptors.FieldDescriptor field : baseUnionDescriptor.getFields()) {
+            Descriptors.Descriptor messageType = fileDescriptor.findMessageTypeByName(field.getMessageType().getName());
+            if (messageType == null) {
+                throw new MetaDataException("Record type " + field.getMessageType().getName() + " removed");
```

Comment:
Yeah agreed that we can do that too. I'm ok either way. Thanks for the review.

Reply:
I think this actually only duplicates the work done by the evolution validator for the moment. (In particular, issue #508 should make it so that the work is only done there.) That work uses the position in the union descriptor as the canonical mapping of old type to new type, whereas this is forced to use the type names. So, I'm okay with this being in both places.